samza-core/src/main/java/org/apache/samza/execution/OperatorSpecGraphAnalyzer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.execution;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import org.apache.samza.operators.OperatorSpecGraph;
+import org.apache.samza.operators.spec.InputOperatorSpec;
+import org.apache.samza.operators.spec.JoinOperatorSpec;
+import org.apache.samza.operators.spec.OperatorSpec;
+
+
+/**
+ * A utility class that encapsulates the logic for traversing an {@link OperatorSpecGraph} and building
+ * associations between related {@link OperatorSpec}s.
+ */
+/* package private */ class OperatorSpecGraphAnalyzer {
```

Comment:
Here are my thoughts on this class. I feel we don't have a pressing need for it to be generified yet.
Can we start with a simple helper class that does traversal and returns a mapping in one go? 

By that, 
* we are still isolating the traversal logic.
* It simplifies the review
* It simplifies testing

We can always refactor to extract the traversal part if we plan to introduce more visitors.
Thoughts?

Reply:
@bharathkk : We'll likely introduce more visitors for computing partition-counts for StreamTableJoin and side-inputs. Once we have those follow-up PRs, we can decide if the current implementation is overly general. If that is indeed the case, we can certainly revisit it.  

Until then, it's probably efficient to leave this PR in its current state. What do you think?

Reply:
If we're only doing stream-stream join then I agree. The reason I wrote it this way though is because I have another PR that will add one more visitor in order to support stream-table joins. I wanted to lay the grounds for the upcoming change since I have already finished coding it.

Reply:
@bharathkk  @vjagadish1989 Would you prefer I send you the follow-up PRs now or is this PR good to go?

Reply:
@ahmedahamid sorry to be late to the party. One comment for your future PR: OperatorSpecGraph is *only* available for high-level APIs. In the effort to unify the runtime support for both high-level and low-level APIs, ExecutionPlanner and the corresponding JobGraph/JobNode classes now have only access to ApplicationDescriptorImpl. Any need to traverse the graph should be starting from ApplicationDescriptorImpl.getInputOperators() now (see PR #642)